File: examples/rust-mobile-demo/android/app/src/main/java/dev/istmo/runtime/Bincode.kt
package dev.istmo.runtime

import java.io.ByteArrayOutputStream

/**
 * Minimal bincode 2 (`standard()` config) codec for the shapes the demo plugin
 * surface needs.
 *
 * Wire format (from `bincode` crate, `VarintEncoding` + `LittleEndian`):
 * * u64 varint — 1 byte for values ≤ 250, otherwise tag byte
 *   (0xFB/0xFC/0xFD) followed by LE u16/u32/u64.
 * * i64 varint — zigzag encoded then u64 varint.
 * * bool — single byte (0x00 / 0x01).
 * * String — u64 varint length + UTF-8 bytes.
 * * Option<T> — 0x00 for None; 0x01 followed by T for Some.
 * * Vec<T> — u64 varint length + T*n.
 * * Tuple `(A, B)` — A followed by B (no length prefix).
 * * enum discriminant — u32 varint.
 */
object Bincode {

    data class Decoded<T>(val value: T, val consumed: Int)

    // ---- Primitives ----

    fun writeBool(out: ByteArrayOutputStream, value: Boolean) {
        out.write(if (value) 1 else 0)
    }

    fun readBool(payload: ByteArray, offset: Int): Decoded<Boolean> {
        require(offset < payload.size) { "bincode bool: buffer underrun" }
        val b = payload[offset].toInt() and 0xFF
        require(b == 0 || b == 1) { "bincode bool: invalid byte 0x${b.toString(16)}" }
        return Decoded(b == 1, offset + 1)
    }

    fun writeString(out: ByteArrayOutputStream, value: String) {
        val utf8 = value.toByteArray(Charsets.UTF_8)
        writeVarintU64(out, utf8.size.toLong())
        out.write(utf8)
    }

    fun writeString(value: String): ByteArray {
        val out = ByteArrayOutputStream()
        writeString(out, value)
        return out.toByteArray()
    }

    /**
     * f32 is written as 4 IEEE 754 little-endian bytes. bincode 2's
     * `standard()` config uses fixed-width encoding for floats — no
     * varint applies to floating-point numbers.
     */
    fun writeF32(out: ByteArrayOutputStream, value: Float) {
        val bits = java.lang.Float.floatToRawIntBits(value)
        for (i in 0 until 4) {
            out.write((bits shr (i * 8)) and 0xFF)
        }
    }

    fun readF32(payload: ByteArray, offset: Int): Decoded<Float> {
        require(offset + 4 <= payload.size) { "bincode f32: buffer underrun" }
        var bits = 0
        for (i in 0 until 4) {
            bits = bits or ((payload[offset + i].toInt() and 0xFF) shl (i * 8))
        }
        return Decoded(java.lang.Float.intBitsToFloat(bits), offset + 4)
    }

    fun readString(payload: ByteArray, offset: Int = 0): Decoded<String> {
        val (length, next) = readVarintU64(payload, offset)
        val end = next + length.toInt()
        require(end <= payload.size) {
            "bincode string overruns buffer: len=$length at offset=$next of ${payload.size}"
        }
        val value = String(payload, next, length.toInt(), Charsets.UTF_8)
        return Decoded(value, end)
    }

    // ---- Combinators ----

    fun <T> writeOption(out: ByteArrayOutputStream, value: T?, writer: (ByteArrayOutputStream, T) -> Unit) {
        if (value == null) {
            out.write(0)
        } else {
            out.write(1)
            writer(out, value)
        }
    }

    fun <T> readOption(payload: ByteArray, offset: Int, reader: (ByteArray, Int) -> Decoded<T>): Decoded<T?> {
        require(offset < payload.size) { "bincode option: buffer underrun" }
        val tag = payload[offset].toInt() and 0xFF
        return when (tag) {
            0 -> Decoded(null, offset + 1)
            1 -> {
                val inner = reader(payload, offset + 1)
                Decoded(inner.value, inner.consumed)
            }
            else -> error("bincode option: invalid tag 0x${tag.toString(16)}")
        }
    }

    fun <T> writeVec(out: ByteArrayOutputStream, items: List<T>, writer: (ByteArrayOutputStream, T) -> Unit) {
        writeVarintU64(out, items.size.toLong())
        for (item in items) writer(out, item)
    }

    fun <T> readVec(payload: ByteArray, offset: Int, reader: (ByteArray, Int) -> Decoded<T>): Decoded<List<T>> {
        val (length, first) = readVarintU64(payload, offset)
        var cursor = first
        val out = ArrayList<T>(length.toInt())
        repeat(length.toInt()) {
            val item = reader(payload, cursor)
            out.add(item.value)
            cursor = item.consumed
        }
        return Decoded(out, cursor)
    }

    fun <A, B> writePair(
        out: ByteArrayOutputStream,
        value: Pair<A, B>,
        writeA: (ByteArrayOutputStream, A) -> Unit,
        writeB: (ByteArrayOutputStream, B) -> Unit,
    ) {
        writeA(out, value.first)
        writeB(out, value.second)
    }

    fun <A, B> readPair(
        payload: ByteArray,
        offset: Int,
        readA: (ByteArray, Int) -> Decoded<A>,
        readB: (ByteArray, Int) -> Decoded<B>,
    ): Decoded<Pair<A, B>> {
        val a = readA(payload, offset)
        val b = readB(payload, a.consumed)
        return Decoded(a.value to b.value, b.consumed)
    }

    // ---- Enums (discriminant only helpers) ----

    /** Write a fieldless enum variant: just the u32-varint discriminant. */
    fun writeEnumDiscriminant(out: ByteArrayOutputStream, discriminant: Int) {
        require(discriminant >= 0) { "enum discriminant must be non-negative" }
        writeVarintU64(out, discriminant.toLong())
    }

    fun readEnumDiscriminant(payload: ByteArray, offset: Int): Decoded<Int> {
        val (value, next) = readVarintU64(payload, offset)
        return Decoded(value.toInt(), next)
    }

    // ---- Varints ----

    fun writeVarintU64(out: ByteArrayOutputStream, value: Long) {
        when {
            value < 0 -> error("bincode varint u64 is unsigned; got $value")
            value <= 250 -> out.write(value.toInt())
            value <= 0xFFFF -> {
                out.write(0xFB)
                writeLittleEndian(out, value, 2)
            }
            value <= 0xFFFF_FFFFL -> {
                out.write(0xFC)
                writeLittleEndian(out, value, 4)
            }
            else -> {
                out.write(0xFD)
                writeLittleEndian(out, value, 8)
            }
        }
    }

    fun readVarintU64(payload: ByteArray, offset: Int): Pair<Long, Int> {
        require(offset < payload.size) { "bincode varint: buffer underrun" }
        val first = payload[offset].toInt() and 0xFF
        return when {
            first <= 250 -> first.toLong() to (offset + 1)
            first == 0xFB -> readLittleEndian(payload, offset + 1, 2) to (offset + 3)
            first == 0xFC -> readLittleEndian(payload, offset + 1, 4) to (offset + 5)
            first == 0xFD -> readLittleEndian(payload, offset + 1, 8) to (offset + 9)
            else -> error("invalid bincode varint prefix: 0x${first.toString(16)}")
        }
    }

    fun writeVarintI64(out: ByteArrayOutputStream, value: Long) {
        // Zigzag encoding then u64 varint.
        val zig = (value shl 1) xor (value shr 63)
        writeVarintU64(out, zig)
    }

    fun readVarintI64(payload: ByteArray, offset: Int): Pair<Long, Int> {
        val (zig, next) = readVarintU64(payload, offset)
        val value = (zig ushr 1) xor -(zig and 1L)
        return value to next
    }

    private fun writeLittleEndian(out: ByteArrayOutputStream, value: Long, bytes: Int) {
        for (i in 0 until bytes) {
            out.write(((value shr (i * 8)) and 0xFF).toInt())
        }
    }

    private fun readLittleEndian(payload: ByteArray, offset: Int, bytes: Int): Long {
        var value = 0L
        for (i in 0 until bytes) {
            value = value or ((payload[offset + i].toLong() and 0xFF) shl (i * 8))
        }
        return value
    }
}

File: examples/data-store-demo/ios/IstmoRuntime/Sources/IstmoRuntime/Bincode.swift
// Minimal port of the bincode 2 "standard" configuration (VarintEncoding +
// LittleEndian) that the istmo wire format uses.
//
// Covers every shape the plugin dispatcher codegen emits:
//
// * varint u8/u16/u32/u64 and i8/i16/i32/i64 (zigzag + varint for signed).
// * varint-length-prefixed String / Data.
// * Bool as a single byte.
// * f32 / f64 as fixed-width IEEE 754 little-endian (bincode 2 keeps
//   floats fixed-width under `VarintEncoding`).
// * Vec / Option combinators with cursor-carrying closures.
//
// Extend per new type as plugin surfaces grow. The generator picks the
// helper it needs from this file; anything the generator names must exist
// here or the Swift compiler surfaces a clear error at the codegen output
// site.

import Foundation

public enum Bincode {

    // MARK: - Writer

    public static func writeBool(_ out: inout Data, _ value: Bool) {
        out.append(value ? 1 : 0)
    }

    public static func writeU8(_ out: inout Data, _ value: UInt8) {
        out.append(value)
    }

    public static func writeI8(_ out: inout Data, _ value: Int8) {
        out.append(UInt8(bitPattern: value))
    }

    /// Bincode 2 "standard" varint encoding for unsigned integers:
    /// values < 251 fit in one byte; larger values are tagged 251/252/253
    /// followed by 2/4/8 little-endian bytes. See `varint::encode_u64` in
    /// the bincode source.
    public static func writeVarintU64(_ out: inout Data, _ value: UInt64) {
        if value < 251 {
            out.append(UInt8(value))
        } else if value <= UInt64(UInt16.max) {
            out.append(251)
            appendLE(&out, UInt16(value))
        } else if value <= UInt64(UInt32.max) {
            out.append(252)
            appendLE(&out, UInt32(value))
        } else {
            out.append(253)
            appendLE(&out, value)
        }
    }

    public static func writeVarintI64(_ out: inout Data, _ value: Int64) {
        writeVarintU64(&out, zigzag(value))
    }

    public static func writeVarintU32(_ out: inout Data, _ value: UInt32) {
        writeVarintU64(&out, UInt64(value))
    }

    public static func writeVarintI32(_ out: inout Data, _ value: Int32) {
        writeVarintI64(&out, Int64(value))
    }

    public static func writeF32(_ out: inout Data, _ value: Float) {
        appendLE(&out, value.bitPattern)
    }

    public static func writeF64(_ out: inout Data, _ value: Double) {
        appendLE(&out, value.bitPattern)
    }

    public static func writeString(_ out: inout Data, _ value: String) {
        let bytes = Array(value.utf8)
        writeVarintU64(&out, UInt64(bytes.count))
        out.append(contentsOf: bytes)
    }

    public static func writeData(_ out: inout Data, _ value: Data) {
        writeVarintU64(&out, UInt64(value.count))
        out.append(value)
    }

    /// Length-prefixed `Vec<T>`. `writer` runs per element and advances
    /// the buffer via inout. `T` is inferred from the closure signature.
    public static func writeVec<T>(
        _ out: inout Data,
        _ items: [T],
        _ writer: (inout Data, T) -> Void
    ) {
        writeVarintU64(&out, UInt64(items.count))
        for item in items {
            writer(&out, item)
        }
    }

    /// One-tag-then-value `Option<T>` encoding.
    public static func writeOption<T>(
        _ out: inout Data,
        _ value: T?,
        _ writer: (inout Data, T) -> Void
    ) {
        if let v = value {
            out.append(1)
            writer(&out, v)
        } else {
            out.append(0)
        }
    }

    // MARK: - Client encode(&payload, value) overloads
    //
    // The client generator emits `Bincode.encode(&payload, arg)` per
    // argument; adding a new arg type in a contract means adding a
    // matching overload here.

    public static func encode(_ out: inout Data, _ value: String) {
        writeString(&out, value)
    }

    public static func encode(_ out: inout Data, _ value: Int32) {
        writeVarintI32(&out, value)
    }

    public static func encode(_ out: inout Data, _ value: Bool) {
        writeBool(&out, value)
    }

    // MARK: - Reader

    public struct Cursor {
        public var data: Data
        public var offset: Int
        public init(_ data: Data) {
            self.data = data
            self.offset = 0
        }
        public var remaining: Int { data.count - offset }
    }

    public enum DecodeError: Error {
        case unexpectedEnd
        case invalidTag(UInt8)
        case invalidUtf8
    }

    public static func readU8(_ c: inout Cursor) throws -> UInt8 {
        guard c.remaining >= 1 else { throw DecodeError.unexpectedEnd }
        let b = c.data[c.offset]
        c.offset += 1
        return b
    }

    public static func readBool(_ c: inout Cursor) throws -> Bool {
        try readU8(&c) != 0
    }

    public static func readVarintU64(_ c: inout Cursor) throws -> UInt64 {
        let tag = try readU8(&c)
        switch tag {
        case 0 ..< 251:
            return UInt64(tag)
        case 251:
            return UInt64(try readLE(&c, UInt16.self))
        case 252:
            return UInt64(try readLE(&c, UInt32.self))
        case 253:
            return try readLE(&c, UInt64.self)
        default:
            throw DecodeError.invalidTag(tag)
        }
    }

    public static func readVarintI64(_ c: inout Cursor) throws -> Int64 {
        unzigzag(try readVarintU64(&c))
    }

    public static func readVarintU32(_ c: inout Cursor) throws -> UInt32 {
        UInt32(try readVarintU64(&c))
    }

    public static func readVarintI32(_ c: inout Cursor) throws -> Int32 {
        Int32(try readVarintI64(&c))
    }

    public static func readF32(_ c: inout Cursor) throws -> Float {
        let bits = try readLE(&c, UInt32.self)
        return Float(bitPattern: bits)
    }

    public static func readF64(_ c: inout Cursor) throws -> Double {
        let bits = try readLE(&c, UInt64.self)
        return Double(bitPattern: bits)
    }

    public static func readString(_ c: inout Cursor) throws -> String {
        let len = Int(try readVarintU64(&c))
        guard c.remaining >= len else { throw DecodeError.unexpectedEnd }
        let slice = c.data.subdata(in: c.offset ..< c.offset + len)
        c.offset += len
        guard let s = String(data: slice, encoding: .utf8) else {
            throw DecodeError.invalidUtf8
        }
        return s
    }

    public static func readData(_ c: inout Cursor) throws -> Data {
        let len = Int(try readVarintU64(&c))
        guard c.remaining >= len else { throw DecodeError.unexpectedEnd }
        let slice = c.data.subdata(in: c.offset ..< c.offset + len)
        c.offset += len
        return slice
    }

    public static func readVec<T>(
        _ c: inout Cursor,
        _ reader: (inout Cursor) throws -> T
    ) throws -> [T] {
        let len = Int(try readVarintU64(&c))
        var out: [T] = []
        out.reserveCapacity(len)
        for _ in 0 ..< len {
            out.append(try reader(&c))
        }
        return out
    }

    public static func readOption<T>(
        _ c: inout Cursor,
        _ reader: (inout Cursor) throws -> T
    ) throws -> T? {
        let tag = try readU8(&c)
        switch tag {
        case 0: return nil
        case 1: return try reader(&c)
        default: throw DecodeError.invalidTag(tag)
        }
    }

    /// Sugar mirroring the `Bincode.decode(bytes) as T` shape the client
    /// generator emits. Currently supports `String`; extend per new type.
    public static func decode(_ bytes: Data) throws -> String {
        var c = Cursor(bytes)
        return try readString(&c)
    }

    // MARK: - Helpers

    private static func zigzag(_ value: Int64) -> UInt64 {
        UInt64(bitPattern: (value << 1) ^ (value >> 63))
    }

    private static func unzigzag(_ value: UInt64) -> Int64 {
        let raw = Int64(bitPattern: value)
        return (raw >> 1) ^ -(raw & 1)
    }

    private static func appendLE(_ out: inout Data, _ value: UInt16) {
        out.append(UInt8(value & 0xff))
        out.append(UInt8((value >> 8) & 0xff))
    }
    private static func appendLE(_ out: inout Data, _ value: UInt32) {
        for i in 0 ..< 4 {
            out.append(UInt8((value >> (i * 8)) & 0xff))
        }
    }
    private static func appendLE(_ out: inout Data, _ value: UInt64) {
        for i in 0 ..< 8 {
            out.append(UInt8((value >> (i * 8)) & 0xff))
        }
    }

    private static func readLE<T: FixedWidthInteger & UnsignedInteger>(
        _ c: inout Cursor, _: T.Type
    ) throws -> T {
        let size = MemoryLayout<T>.size
        guard c.remaining >= size else { throw DecodeError.unexpectedEnd }
        var out: T = 0
        for i in 0 ..< size {
            out |= T(c.data[c.offset + i]) << (i * 8)
        }
        c.offset += size
        return out
    }
}
